Upgrade topbar extensions to JupyterLab 4

mahendrapaipuri opened the pull request:
- This PR primarily works towards upgrading the topbar extensions to JupyterLab 4 using the default [topbar](https://github.com/jupyterlab/jupyterlab/blob/main/packages/application-extension/src/topbar.ts) extension provided by JupyterLab.
- Relevant topbar extensions, [jupyterlab-topbar-text](https://github.com/jupyterlab-contrib/jupyterlab-topbar-text/), [jupyterlab-logout](https://github.com/jupyterlab-contrib/jupyterlab-logout), [jupyterlab-theme-toggler](https://github.com/jtpio/jupyterlab-theme-toggle), [jupyterlab-system-monitor](https://github.com/jtpio/jupyterlab-system-monitor) are moved into a monorepo to ease the maintenance burden
- Repo organisation has been inspired from [jupyter-renderers](https://github.com/jupyterlab/jupyter-renderers)
- GitHub workflows have been modernised to use [jupyterlab-maintainer-tools](https://github.com/jupyterlab/maintainer-tools) actions.

Closes #81

jtpio commented:
Nice thanks @mahendrapaipuri for working on this!

jtpio commented on .stylelintcache:
Looks like this file can be added to the `.gitignore`

mahendrapaipuri replied:
Yeah, totally. Cheers for pointing it out!!

jtpio commented on README.md:
```diff
@@ -15,11 +15,10 @@ Inspired by Gnome Shell Top Bar indicators.
-- [jupyterlab-logout](https://github.com/jupyterlab-contrib/jupyterlab-logout): add a "Log Out" button
-- [jupyterlab-theme-toggle](https://github.com/jtpio/jupyterlab-theme-toggle): switch between the Light and Dark themes
+- [jupyterlab-topbar-text](./packages/topbar-text-extension): add and edit custom text
+- [jupyterlab-system-monitor](./packages/system-monitor-extension): show system metrics (memory usage)
```
For `jupyterlab-system-monitor` either we include it here, or maybe directly in `jupyter-resource-usage`? https://github.com/jupyter-server/jupyter-resource-usage/issues/191

mahendrapaipuri replied:
I agree that it makes more sense to move it over to `jupyter-resource-usage` as users can install everything at once. I will work on that!!

jtpio replied:
that would be great, thanks!

jtpio replied:
For reference there is a PR to update `jupyter-resource-usage` to JupyterLab 4: https://github.com/jupyter-server/jupyter-resource-usage/pull/195, that could be finished before https://github.com/jupyter-server/jupyter-resource-usage/issues/191.

Or update to JupyterLab 4 directly as part of https://github.com/jupyter-server/jupyter-resource-usage/issues/191.

mahendrapaipuri replied:
@jtpio How do we organise the `jupyter-resource-usage`? Should I wait for you to finish your [PR](https://github.com/jupyter-server/jupyter-resource-usage/pull/195) and then add the lab extension package?

jtpio replied:
I'll have very low bandwidth over the next 2 weeks so if you want to take over and finish the lab 4 PR that would be greatly appreciated. Thanks!

mahendrapaipuri replied:
Sure, I can do that. Are you going to add me as a collaborator or should I start a new PR with your branch as base?

jtpio commented on packages/theme-toggler-extension/jupyterlab-theme-toggler/_version.py:
```diff
@@ -0,0 +1,4 @@
+# This file is auto-generated by Hatchling. As such, do not:
+#   - modify
+#   - track in version control e.g. be sure to add to .gitignore
+__version__ = VERSION = '1.0.0'
```
+1 for going with `1.0.0` for all these extensions :+1:

mahendrapaipuri replied:
And `_version.py` must be ignored as well. Will fix these soon.

jtpio commented:
Thanks again @mahendrapaipuri for working on this.

I updated the repo settings so the CI should run when new changes are pushed, to reduce friction.

mahendrapaipuri commented:
@jtpio Fixed issues with CI. This PR is good to go for the review.

I dont have an awful lot of experience with GitHub workflows. I tried to modernise the actual CI workflows without changing the way the packages are being tested. If you think, we can improve/optimize the existing workflows, please let me know.

jtpio commented on packages/logout-extension/src/index.ts:
```diff
@@ -0,0 +1,35 @@
+    // Get app commands
+    const { commands } = app;
+
+    const namespace = 'topbarcmds';
```
For consistency with other third-party extensions, should we rename the namespace to `jupyterlab-logout`?

```suggestion
    const namespace = 'jupyterlab-logout';
```

mahendrapaipuri replied:
What do you think about `jupyterlab-topbar` as namespace? It will still be project specific namespace and we will be able to use same namespace for any future buttons we will add to topbar.

jtpio replied:
That could indeed work, unless we consider all these extensions as independent and the first part of the command (namespace) to correspond to the name of the extension.

jtpio commented on packages/logout-extension/src/index.ts:
```diff
@@ -0,0 +1,35 @@
+
+import '../style/index.css';
+
+const logoutPluginId = '@jupyterlab/logout-extension:plugin';
```
Since this is not an official extension, maybe the following name would be more appropriate?

```suggestion
const logoutPluginId = 'jupyterlab-logout-extension:plugin';
```

jtpio replied:
For consistency we could then do the same with the other extensions in this repo / diff.

mahendrapaipuri replied:
Fair enough. I thought this convention was put in-place when extensions were distributed as npm packages to put all official extensions under one umbrella namespace. As this is not the case anymore, I thought we could use this convention so that all extensions will be installed at same location for the end users. 

Anyways, I will change the names for all extensions.

jtpio commented on packages/logout-extension/package.json:
```diff
@@ -0,0 +1,84 @@
+{
+  "name": "jupyterlab-logout-extension",
```
Should the `npm` name keep the `-extension` suffix? Or just be `jupyterlab-logout` like the current name on npm? https://www.npmjs.com/package/jupyterlab-logout

jtpio replied:
Just so that when we make a release later with this change we still publish to the same package on npm.

jtpio commented on packages/theme-toggler-extension/package.json:
```diff
@@ -0,0 +1,87 @@
+{
+  "name": "jupyterlab-theme-toggler-extension",
```
Similar comment as above, maybe we should stick to `jupyterlab-theme-toggle` for now?

https://www.npmjs.com/package/jupyterlab-theme-toggle

jtpio commented:
Thanks @mahendrapaipuri for your work on this :+1: 

I left a few more comments about how to name the npm packages. Looks like the PR should be good to go. We could also merge and iterate in follow-up PRs.

One thing to check in a separate PR would be to add support for the [Jupyter Releaser](https://jupyter-releaser.readthedocs.io/en/latest/) so we can automate the publishing of all the Python packages (and npm) in the monorepo.

mahendrapaipuri commented:
Happy to contribute and thanks to you too @jtpio for helping me out.

Addressed all your PR comments. Yes, I think we can add support for releaser in a separate PR. 

I see that you merged [#195](https://github.com/jupyter-server/jupyter-resource-usage/pull/195) that adds JLab 4 support for `jupyter-resource-usage`. I will put up a PR with `jupyterlab-resource-usage` extension added to that project.

jtpio reviewed (approved):
Thanks!

This looks good to go. As mentioned in the comments we'll have to check how to setup the versioning of the extensions so releases can be made with the Jupyter Releaser.

jtpio commented:
> I see that you merged [#195](https://github.com/jupyter-server/jupyter-resource-usage/pull/195) that adds JLab 4 support for `jupyter-resource-usage`. I will put up a PR with `jupyterlab-resource-usage` extension added to that project.

Yes that would be great, thanks!
